File: parts/src/org/lineageos/settings/BootCompletedReceiver.java
/*
 * Copyright (C) 2015 The CyanogenMod Project
 *               2017-2019 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.lineageos.settings;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.hardware.display.DisplayManager;
import android.os.IBinder;
import android.util.Log;
import android.view.Display;
import android.view.Display.HdrCapabilities;

import org.lineageos.settings.doze.DozeUtils;
import org.lineageos.settings.thermal.ThermalUtils;
import org.lineageos.settings.refreshrate.RefreshUtils;

public class BootCompletedReceiver extends BroadcastReceiver {
    private static final boolean DEBUG = false;
    private static final String TAG = "XiaomiParts";

    @Override
    public void onReceive(final Context context, Intent intent) {
        if (DEBUG)
            Log.d(TAG, "Received boot completed intent");
        DozeUtils.onBootCompleted(context);
        ThermalUtils.startService(context);
        RefreshUtils.startService(context);
        overrideHdrTypes(context);
    }

    private static void overrideHdrTypes(Context context) {
        // Override HDR types to enable Dolby Vision
        final DisplayManager dm = context.getSystemService(DisplayManager.class);
        dm.overrideHdrTypes(Display.DEFAULT_DISPLAY, new int[]{
                HdrCapabilities.HDR_TYPE_DOLBY_VISION, HdrCapabilities.HDR_TYPE_HDR10,
                HdrCapabilities.HDR_TYPE_HLG, HdrCapabilities.HDR_TYPE_HDR10_PLUS});
    }
}

File: parts/src/co/infinity/xparts/services/BootCompletedReceiver.kt
/*
 * Copyright (C) 2015 The CyanogenMod Project
 * SPDX-FileCopyrightText: Paranoid Android
 * SPDX-FileCopyrightText: Project Infinity X
 * SPDX-License-Identifier: Apache-2.0
 */

package co.infinity.xparts.services

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.hardware.display.DisplayManager
import android.view.Display
import android.view.Display.HdrCapabilities
import co.infinity.xparts.data.AuxCameraUtils
import co.infinity.xparts.data.FpGestureUtils
import co.infinity.xparts.data.SaturationUtils
import co.infinity.xparts.data.ThermalUtils
import co.infinity.xparts.utils.Logging

class BootCompletedReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        Logging.i(TAG, "Received intent: ${intent.action}")
        
        when (intent.action) {
            Intent.ACTION_LOCKED_BOOT_COMPLETED -> handleLockedBootCompleted(context)
            Intent.ACTION_BOOT_COMPLETED -> handleBootCompleted(context)
        }
    }

    private fun handleLockedBootCompleted(context: Context) {
        Logging.i(TAG, "Handling locked boot completed.")
        try {
            // 1. Initialize Services
            startServices(context)

            // 2. Override HDR Types
            overrideHdrTypes(context)
            
            // 3. Restore Saturation
            SaturationUtils.getInstance(context).restoreSaturation()

        } catch (e: Exception) {
            Logging.e(TAG, "Error during locked boot completed processing", e)
        }
    }

    private fun handleBootCompleted(context: Context) {
        Logging.i(TAG, "Handling boot completed.")
        // Add additional boot-completed actions if needed
    }

    private fun startServices(context: Context) {
        Logging.i(TAG, "Starting services...")
	
        // Aux Camera Services
        AuxCameraUtils.restore(context)

        // Thermal Services
        ThermalUtils.getInstance(context).startService()

        // Fingerprint Gestures
        if (FpGestureUtils.isAvailable(context) && FpGestureUtils.isEnabled(context)) {
            Logging.d(TAG, "Enabling Fingerprint Gesture on boot")
            FpGestureUtils.updateDriver(true)
        }
    }

    private fun overrideHdrTypes(context: Context) {
        try {
            val dm = context.getSystemService(DisplayManager::class.java)
            if (dm != null) {
                dm.overrideHdrTypes(
                    Display.DEFAULT_DISPLAY, 
                    intArrayOf(
                        HdrCapabilities.HDR_TYPE_DOLBY_VISION,
                        HdrCapabilities.HDR_TYPE_HDR10,
                        HdrCapabilities.HDR_TYPE_HLG,
                        HdrCapabilities.HDR_TYPE_HDR10_PLUS
                    )
                )
                Logging.i(TAG, "HDR types overridden successfully.")
            }
        } catch (e: Exception) {
            Logging.e(TAG, "Error overriding HDR types", e)
        }
    }

    companion object {
        private const val TAG = "BootCompletedReceiver"
    }
}
